File: frontend/src/api/orchestration/providers.ts
import { apiFetch } from "../client";

export type ProviderConfig = {
    id: string;
    project_id: string | null;
    name: string;
    provider_type: string;
    base_url: string | null;
    api_key_hint: string | null;
    organization: string | null;
    default_model: string;
    fallback_model: string | null;
    temperature: number;
    max_tokens: number;
    timeout_seconds: number;
    is_default: boolean;
    is_enabled: boolean;
    metadata: Record<string, unknown>;
    last_healthcheck_status: string | null;
    last_healthcheck_latency_ms: number | null;
    is_healthy: boolean;
    last_healthcheck_at: string | null;
    created_at: string;
    updated_at: string;
};

export type ProviderModelList = {
    provider_id: string;
    provider_type: string;
    models: Array<Record<string, unknown>>;
};

export type ModelCapability = {
    id: string;
    provider_id: string | null;
    provider_type: string;
    model_slug: string;
    display_name: string | null;
    supports_tools: boolean;
    supports_tool_calling: boolean;
    supports_structured_output: boolean;
    supports_reasoning: boolean;
    supports_vision: boolean;
    max_context_tokens: number;
    cost_per_1k_input: number;
    cost_per_1k_output: number;
    context_window: number | null;
    max_output_tokens: number | null;
    input_cost_per_1k: number | null;
    output_cost_per_1k: number | null;
    input_cost_per_1m: number | null;
    output_cost_per_1m: number | null;
    latency_p50: number | null;
    health_status: string | null;
    source_for_each_field: Record<string, string>;
    last_verified_at: string | null;
    override_reason: string | null;
    metadata: Record<string, unknown>;
    is_active: boolean;
    created_at: string;
    updated_at: string;
};

export type ProviderCompareResult = {
    provider_id: string;
    provider_name: string;
    provider_type: string;
    model_name: string;
    latency_ms: number;
    input_tokens: number;
    output_tokens: number;
    token_total: number;
    estimated_cost_usd: number;
    output_text: string;
    is_healthy: boolean;
};

export type ProviderCompareResponse = {
    prompt_preview: string;
    result_a: ProviderCompareResult;
    result_b: ProviderCompareResult;
};

export type ProviderHealthSummary = {
    provider_id: string;
    project_id: string | null;
    name: string;
    provider_type: string;
    default_model: string;
    enabled: boolean;
    status: string;
    healthy: boolean | null;
    latency_ms: number | null;
    last_checked_at: string | null;
    error: string | null;
};

export async function listProviders(projectId?: string): Promise<ProviderConfig[]> {
    const suffix = projectId ? `?project_id=${encodeURIComponent(projectId)}` : "";
    return apiFetch(`/orchestration/providers${suffix}`);
}

export async function createProvider(payload: Record<string, unknown>): Promise<ProviderConfig> {
    return apiFetch("/orchestration/providers", { method: "POST", body: JSON.stringify(payload) });
}

export async function updateProvider(providerId: string, payload: Record<string, unknown>): Promise<ProviderConfig> {
    return apiFetch(`/orchestration/providers/${providerId}`, { method: "PATCH", body: JSON.stringify(payload) });
}

export async function deleteProvider(providerId: string): Promise<void> {
    await apiFetch(`/orchestration/providers/${providerId}`, { method: "DELETE" });
}

export async function testProvider(providerId: string): Promise<Record<string, unknown>> {
    return apiFetch(`/orchestration/providers/${providerId}/test`, { method: "POST" });
}

export async function runProviderHealthChecks(): Promise<Array<Record<string, unknown>>> {
    return apiFetch("/orchestration/providers/health-check", { method: "POST" });
}

export async function getProviderHealthSummary(): Promise<ProviderHealthSummary[]> {
    return apiFetch("/orchestration/providers/health-summary");
}

export async function startProviderRuntime(providerId: string): Promise<Record<string, unknown>> {
    return apiFetch(`/orchestration/providers/${providerId}/runtime/start`, { method: "POST" });
}

export async function listProviderModels(providerId: string): Promise<ProviderModelList> {
    return apiFetch(`/orchestration/providers/${providerId}/models`);
}

export async function listModelCapabilities(): Promise<ModelCapability[]> {
    return apiFetch("/orchestration/providers/model-capabilities");
}

export async function compareProviders(payload: Record<string, unknown>): Promise<ProviderCompareResponse> {
    return apiFetch("/orchestration/providers/compare", {
        method: "POST",
        body: JSON.stringify(payload),
    });
}

File: backend/modules/orchestration/schemas/providers.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

from backend.core.schemas import RequestModel
from backend.modules.github.schemas import GithubSyncEventResponse

from backend.modules.orchestration.schemas.common import *  # noqa: F403

class ProviderConfigCreate(RequestModel):
    project_id: str | None = None
    name: str = Field(min_length=2, max_length=255)
    provider_type: str = Field(min_length=2, max_length=64)
    base_url: str | None = None
    api_key: str | None = None
    organization: str | None = None
    default_model: str = Field(min_length=1, max_length=255)
    fallback_model: str | None = None
    temperature: float = Field(default=0.2, ge=0, le=2)
    max_tokens: int = Field(default=4096, ge=128, le=200000)
    timeout_seconds: int = Field(default=120, ge=5, le=3600)
    is_default: bool = False
    is_enabled: bool = True
    metadata: dict[str, Any] = Field(default_factory=dict)


class ProviderConfigUpdate(RequestModel):
    name: str | None = None
    base_url: str | None = None
    api_key: str | None = None
    organization: str | None = None
    default_model: str | None = None
    fallback_model: str | None = None
    temperature: float | None = Field(default=None, ge=0, le=2)
    max_tokens: int | None = Field(default=None, ge=128, le=200000)
    timeout_seconds: int | None = Field(default=None, ge=5, le=3600)
    is_default: bool | None = None
    is_enabled: bool | None = None
    metadata: dict[str, Any] | None = None


class ProviderConfigResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    project_id: str | None
    name: str
    provider_type: str
    base_url: str | None
    api_key_hint: str | None
    organization: str | None
    default_model: str
    fallback_model: str | None
    temperature: float
    max_tokens: int
    timeout_seconds: int
    is_default: bool
    is_enabled: bool
    metadata: dict[str, Any]
    last_healthcheck_status: str | None
    last_healthcheck_latency_ms: int | None
    is_healthy: bool
    last_healthcheck_at: datetime | None
    created_at: datetime
    updated_at: datetime


class ModelCapabilityResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    provider_id: str | None
    provider_type: str
    model_slug: str
    display_name: str | None
    supports_tools: bool
    supports_tool_calling: bool = False
    supports_structured_output: bool = False
    supports_reasoning: bool = False
    supports_vision: bool
    max_context_tokens: int
    cost_per_1k_input: float
    cost_per_1k_output: float
    context_window: int | None = None
    max_output_tokens: int | None = None
    input_cost_per_1k: float | None = None
    output_cost_per_1k: float | None = None
    input_cost_per_1m: float | None = None
    output_cost_per_1m: float | None = None
    latency_p50: int | None = None
    health_status: str | None = None
    source_for_each_field: dict[str, str] = Field(default_factory=dict)
    last_verified_at: datetime | None = None
    override_reason: str | None = None
    metadata: dict[str, Any]
    is_active: bool
    created_at: datetime
    updated_at: datetime


class ProviderModelListResponse(BaseModel):
    provider_id: str
    provider_type: str
    models: list[dict[str, Any]] = Field(default_factory=list)


class ProviderCompareRequest(RequestModel):
    provider_a_id: str
    provider_b_id: str
    model_a: str | None = None
    model_b: str | None = None
    task_title: str = Field(min_length=2, max_length=255)
    task_description: str | None = None
    acceptance_criteria: str | None = None
    task_metadata: dict[str, Any] = Field(default_factory=dict)


class ProviderCompareResult(BaseModel):
    provider_id: str
    provider_name: str
    provider_type: str
    model_name: str
    latency_ms: int
    input_tokens: int
    output_tokens: int
    token_total: int
    estimated_cost_usd: float
    output_text: str
    is_healthy: bool


class ProviderCompareResponse(BaseModel):
    prompt_preview: str
    result_a: ProviderCompareResult
    result_b: ProviderCompareResult


